File: Analysis and Design of Algorithms/Practice 2/bfs.cpp
#include <iostream>
#include <conio.h>
#include <vector>
#include <list>
#include <queue>

using namespace std;

class Graph
{
    int V;
    list<int> *l;

public:
    Graph(int v)
    {
        V = v;
        l = new list<int>[V];
    }

    void addEdge(int u, int v, bool undir = true)
    {
        l[u].push_back(v);
        if (undir)
            l[v].push_back(u);
    }

    void bfs(int src)
    {
        vector<int> ans;
        vector<bool> visited(V, false);
        queue<int> q;

        q.push(src);
        visited[src] = true;

        while (!q.empty())
        {
            int f = q.front();
            ans.push_back(f);
            q.pop();

            for (int nbr : l[f])
            {
                if (!visited[nbr])
                {
                    visited[nbr] = true;
                    q.push(nbr);
                }
            }
        }

        for (int nbr : ans)
            cout << nbr << ", ";
    }
};

int main()
{
    Graph g(7);

    g.addEdge(1, 2);
    g.addEdge(1, 3);
    g.addEdge(2, 4);
    g.addEdge(2, 5);
    g.addEdge(3, 5);
    g.addEdge(4, 5);
    g.addEdge(4, 6);
    g.addEdge(5, 6);

    cout << "Traversal order is: ";

    cout << "\nUsing BFS => ";
    g.bfs(1);

    // cout << "\nUsing DFS =>";
    // g.dfs(1);

    getch();
    return 0;
}

File: Artificial Intelligence/Practice/bfs.py
class graph:
    vertices = {}

    def __init__(self):
        pass

    def addEdge(self, u, v):
        if u in self.vertices:
            self.vertices[u].append(v)
        else:
            l = list()
            l.append(v)
            self.vertices[u] = l

    def printGraph(self):
        for i in self.vertices:
            print(f"{i} -> {self.vertices[i]}")

    def bfs(self, start, goal):
        visited = set()
        queue = []

        if(start == goal):
            print("Goal == Start")
            return

        visited.add(start)
        queue.append(start)

        while(len(queue) != 0):
            front = queue.pop(0)
            print(f"{front}", end=" ")

            if(front == goal):
                print("\nGoal Found!")
                return

            try:
                for n in self.vertices[front]:
                    if(n not in visited):
                        queue.append(n)
                        visited.add(n)
            except:
                pass

        print("Traversal Complete. Goal not found!")


g = graph()
e = 7

# print("Enter the edges: ")
# for i in range(e):
#     edge = input()
#     edge.split()
#     g.addEdge(int(edge[0]), int(edge[2]))

g.addEdge(0, 1)
g.addEdge(0, 2)
g.addEdge(1, 3)
g.addEdge(3, 4)
g.addEdge(4, 0)
g.addEdge(4, 1)
g.addEdge(4, 5)

print("\nGraph is:")
g.printGraph()

start = int(input("\nEnter the start node: "))
goal = int(input("Enter the goal node: "))

print("\nBFS Traversal is: ", end="")
g.bfs(start, goal)
